### ISSUE TITLE
Report unused dependencies in CI

### ISSUE THREAD

**Issue by rnveach** [checkstyle/checkstyle#11065]
Thought about at https://github.com/checkstyle/checkstyle/pull/11064 and then identified at https://www.baeldung.com/maven-unused-dependencies ,

We need a way to identify unused dependencies and be alerted that they should be removed.

We should add `mvn dependency:analyze` to our CI to report unused dependencies.

This may not get much usage, but I think it would be good to make sure we don't have to worry about it anymore too.

**Comment by rnveach** [comment]
Running this locally on master provided the following output:
````
[WARNING] Used undeclared dependencies found:
[WARNING]    com.google.code.findbugs:jsr305:jar:3.0.2:compile
[WARNING]    com.sun:tools:jar:1.8.0:system
[WARNING]    org.junit.jupiter:junit-jupiter-params:jar:5.7.2:test
[WARNING]    org.mockito:mockito-core:jar:4.1.0:compile
[WARNING]    org.junit.platform:junit-platform-commons:jar:1.8.2:test
[WARNING]    net.bytebuddy:byte-buddy:jar:1.12.1:compile
[WARNING]    junit:junit:jar:4.13.2:test
[WARNING]    org.hamcrest:hamcrest-core:jar:1.3:test
[WARNING]    commons-logging:commons-logging:jar:1.2:compile
[WARNING]    com.tngtech.archunit:archunit:jar:0.22.0:test
[WARNING] Unused declared dependencies found:
[WARNING]    com.tngtech.archunit:archunit-junit5:jar:0.22.0:test
[WARNING]    org.junit.jupiter:junit-jupiter-engine:jar:5.8.2:test
[WARNING]    org.junit.vintage:junit-vintage-engine:jar:5.8.2:test
[WARNING]    com.github.stefanbirkner:system-rules:jar:1.19.0:test
[WARNING]    org.mockito:mockito-inline:jar:4.1.0:compile
[WARNING]    org.slf4j:slf4j-simple:jar:1.7.32:test
[WARNING]    org.jacoco:org.jacoco.agent:jar:runtime:0.8.7:test
[WARNING] Non-test scoped test only dependencies found:
[WARNING]    com.google.code.findbugs:jsr305:jar:3.0.2:compile
[WARNING]    com.google.guava:guava:jar:31.0.1-jre:compile
[WARNING]    org.mockito:mockito-core:jar:4.1.0:compile
[WARNING]    net.bytebuddy:byte-buddy:jar:1.12.1:compile
[WARNING]    info.picocli:picocli:jar:4.6.2:compile
````

**Comment by romani** [comment]
I do not think we can remove any of this dependencies.

but 
https://github.com/checkstyle/checkstyle/blob/2a4226995f89d73b675e5209ffba7431c438d4fe/pom.xml#L318-L322
is good catch ... why it is without test scope, @pbludov  ?

**Comment by romani** [comment]
byte-buddy is not our declared dependency. It might be transitive.
picocli is used by our CLI class a tool to do fancy CLI support.

output of this plugin  is not very clear how to interpretate.

**Comment by rnveach** [comment]
> mockito-inline

Even with test scope, this dependency still shows up. Code fails to compile without it.

> output of this plugin is not very clear how to interpretate.

Agree. from what I have seen false positives seem possible.

**Comment by romani** [comment]
if amount of false positives is too much, we should think twice to use it, it will bother us a lot in maintenance.

**Comment by rnveach** [comment]
Plugin allows suppressions. Example:
````
<plugin>
    <groupId>org.apache.maven.plugins</groupId>
    <artifactId>maven-dependency-plugin</artifactId>
    <configuration>
        <usedDependencies>
            <dependency>commons-collections:commons-collections</dependency>
        </usedDependencies>
    </configuration>
</plugin>
````

Additional information:
> sometimes we might need to load classes at runtime, like for example in a plugin oriented project.
> Since the dependencies are not used at compile-time, the maven-dependency-plugin would issue a warning stating that a dependency is not being used, when in fact it is.

**Comment by romani** [comment]
@pbludov , @nmancus1 , @strkkk , please share your vote to use this plugin or not ?

for now it caught two probems: missed test scope in mockito, and junit-vintage-engine.

**Comment by rnveach** [comment]
Guava is being reported as `Non-test scoped test only dependencies found`. I connected this to an open issue in their issue tracker at https://issues.apache.org/jira/browse/MDEP-753 . Any suppressions added should probably go under this issue.

**Comment by rnveach** [comment]
https://stackoverflow.com/questions/4565740/what-are-unused-undeclared-dependencies-in-maven-what-to-do-with-them

> Used undeclared dependencies are those which are required, but have not been explicitly declared as dependencies in your project. They are however available thanks to transitive dependency of other dependencies in your project. It is a good idea to explicitly declare these dependencies. This also allows you to control the version of these dependencies (perhaps matching the version provided by your runtime).

>  unused declared dependencies, it is a good idea to remove them. Why add unnecessary dependency to your project? But then transitivity can bring these in anyway, perhaps, conflicting with your runtime versions. In this case, you will need to specify them — essentially to control the version.

**Comment by nmancus1** [comment]
> please share your vote to use this plugin or not

Let's try it, and if exclusions/ updates become annoying, we can revisit usage.

**Comment by romani** [comment]
during release I saw:
```
[INFO] Including net.sf.saxon:Saxon-HE:jar:10.6 in the shaded jar.
[WARNING] Saxon-HE-10.6.jar, antlr4-runtime-4.9.3.jar, checker-qual-3.12.0.jar, checkstyle-9.2.1.jar, commons-beanutils-1.9.4.jar, commons-collections-3.2.2.jar, commons-logging-1.2.jar, error_prone_annotations-2.7.1.jar, failureaccess-1.0.1.jar, guava-31.0.1-jre.jar, j2objc-annotations-1.3.jar, javassist-3.28.0-GA.jar, jsr305-3.0.2.jar, listenablefuture-9999.0-empty-to-avoid-conflict-with-guava.jar, picocli-4.6.2.jar, reflections-0.10.2.jar, slf4j-api-1.7.32.jar define 1 overlapping resource: 
[WARNING]   - META-INF/MANIFEST.MF
[WARNING] checker-qual-3.12.0.jar, commons-beanutils-1.9.4.jar, commons-collections-3.2.2.jar, commons-logging-1.2.jar define 1 overlapping resource: 
[WARNING]   - META-INF/LICENSE.txt
[WARNING] commons-beanutils-1.9.4.jar, commons-collections-3.2.2.jar, commons-logging-1.2.jar define 1 overlapping resource: 
[WARNING]   - META-INF/NOTICE.txt
[WARNING] maven-shade-plugin has detected that some class files are
[WARNING] present in two or more JARs. When this happens, only one
[WARNING] single version of the class is copied to the uber jar.
[WARNING] Usually this is not harmful and you can skip these warnings,
[WARNING] otherwise try to manually exclude artifacts based on
[WARNING] mvn dependency:tree -Ddetail=true and the above output.
[WARNING] See http://maven.apache.org/plugins/maven-shade-plugin/
[INFO] Attaching shaded artifact.

```

**Comment by pbludov** [comment]
SO [suggests](https://stackoverflow.com/questions/22968266/maven-shade-plugin-and-licenses) to use [ApacheLicenseResourceTransformer](https://maven.apache.org/plugins/maven-shade-plugin/examples/resource-transformers.html#ApacheLicenseResourceTransformer) to keep all LICENSEs included into the uber jar.

**Comment by romani** [comment]
![Screenshot_20220106-220151](https://user-images.githubusercontent.com/812984/148499552-e20c7aeb-30b5-4c3e-bd6e-75d8d527d909.png)


Why we got +1Mb ?

**Comment by rnveach** [comment]
> Why we got +1Mb ?

A new folder, `javaassit`, was added that resulted in 820kb of files.

**Comment by romani** [comment]
Eclipse Plugin experienced problems with missed such dependency
https://github.com/checkstyle/eclipse-cs/pull/336#issuecomment-1007159696

**Comment by rnveach** [comment]
Yes, after sleeping on it, I find it weird we were excluding a production bound library and figured if it was needed it would have repercussions down the line.

**Comment by Calixte** [comment]
javassist is needed at runtime by [MetadataFactory](https://github.com/checkstyle/eclipse-cs/blob/master/net.sf.eclipsecs.core/src/net/sf/eclipsecs/core/config/meta/MetadataFactory.java) after the upgrade to org.reflections 0.10.x ; it's probably also needed by [XmlMetaReader](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/meta/XmlMetaReader.java) which is only used by plugins.
It would be nice if it kept shipping with the checkstyle-all jar (like it started doing in 9.2.1).
